utils/pennsylvania.py

```diff
@@ -0,0 +1,643 @@
+        return df
+
+    def merge_contrib_filer_datasets(
+        self, cont_file: pd.DataFrame, filer_file: pd.DataFrame
+    ) -> pd.DataFrame:
+        """merges the contributor and filer datasets from the same year using the
+        unique filerID
+        Args:
+            cont_file: The contributor dataset
+
+            filer_file: the filer dataset from the same year as the contributor
+            file.
+        Returns
+            The merged pandas dataframe
+        """
+        merged_df = pd.merge(cont_file, filer_file, how="left", on="RECIPIENT_ID")
+        return merged_df
+
+    def merge_expend_filer_datasets(
+        self, expend_file: pd.DataFrame, filer_file: pd.DataFrame
+    ) -> pd.DataFrame:
+        """merges the expenditure and filer datasets from the same year using the
+        unique filerID
+        Args:
+            expend_file: The expenditure dataset
+
+            filer_file: the filer dataset from the same year as the expenditure file
+        Returns
+            The merged pandas dataframe
+        """
+        merged_df = pd.merge(
+            expend_file, filer_file, left_on="DONOR_ID", right_on="RECIPIENT_ID"
+        ).drop("RECIPIENT_ID", axis=1)
+        return merged_df
+
+    def format_contrib_data_for_concat(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Reformartes the merged contributor-filer dataset such that it has the
+        same columns as the merged expenditure-filer dataset so that concatenation
+        can occur
+
+        Args:
+            The merged contributor-filer dataset
+
+        Returns:
+            A new dataframe with the appropriate column formatting for concatenation
+        """
+        df["DONOR_ID"] = None
+        df["DONOR_PARTY"] = None
+        df["DONOR_OFFICE"] = None
+        columns = df.columns.to_list()
+        columns.sort()
+        df = df.loc[:, columns]
+        return df
+
+    def format_expend_data_for_concat(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Reformartes the merged expenditure-filer dataset such that it has the
+        same columns as the merged contributor-filer dataset so that concatenation
+        can occur
+
+        Args:
+            The merged expenditure-filer dataset
+
+        Returns:
+            A new dataframe with the appropriate column formatting for concatenation
+        """
+        df["RECIPIENT_ID"] = None
+        df = df.rename(
+            columns={
+                "RECIPIENT_x": "RECIPIENT",
+                "RECIPIENT_y": "DONOR",
+                "RECIPIENT_TYPE": "DONOR_TYPE",
+                "RECIPIENT_PARTY": "DONOR_PARTY",
+                "RECIPIENT_OFFICE": "DONOR_OFFICE",
+            }
+        )
+        # because recipient information in the expenditure dataset is not
+        # provided, for the sake of fitting the schema I code recipient_type as
+        # 'transaction'.
+        df["RECIPIENT_TYPE"] = "Transaction"
```

Comment:
WHAT????? @alankagiri

Reply:
In the transactions table we said that for those people whose 'entity_type' isn't specified, like my expenditure data, we code 'transaction'. This is what this is doing, and it would only occur for data coming from expenditure data. Isn't that what we agreed on?